Add static UIJob-factories similiar to Job factories

HannesWell opened the pull request:
The Job class has convenient factory methods `create()` and `createSystem()` to create a job that executes a given `ICoreRunnable` or `IJobFunction`.
This PR proposes to add similar factories for the subclass `UIJob` named `UIJob.createUI()`. The name `createUI` was chosen to not clash with the super-class.

HannesWell committed:
Add static UIJob-factories similar to Job factories

Additionally handle cancellation by throwing OperationCanceledException
correctly.

HannesWell commented on bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/progress/UIJob.java:
```diff
@@ -31,6 +34,53 @@
+	 * Creates a new UIJob that will execute the provided function in the UI thread
+	 * when it runs.
+	 *
+	 * Prefer using {@link UIJob#createUI(String, ICoreRunnable)}.
```
The Job class recommends to prefer the ICoreRunnable overload, but the mentioned reason is not valid anymore because both get monitor.done() called and can use `OperationCanceledException`.

Therefore I wonder if the recommendation is still required or if just the `ICoreRunnable` should be added?

HannesWell replied:
Since we have a factory with `IJobFunction` and `ICoreRunnable`, the method calls is ambiguous for non-block lambda expressions, which is a bit ugly because then one has to cast the expression to `IJobFunction` or `ICoreRunnable`.
If I only add a `ICoreRunnable` that problem wouldn't occur.
Correct me if I'm wrong, but I have the impression that `ICoreRunnable` is the more modern approach and the `IJobFunction`  factory is only in the Job class for backwards compatibility and one can achieve the same regarding error-propagation and cancellation with an ICoreRunnable. Therefore I'm tending more towards just adding the ICoreRunnable factory.

iloveeclipse commented on bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/progress/UIJob.java:
```diff
@@ -31,6 +34,53 @@
+	 * @see IJobFunction
+	 * @since 3.127
+	 */
+	public static UIJob createUI(String name, IJobFunction function) {
```
why "createUI" and not simply "create"? We don't create widgets here :)

HannesWell replied:
> why "createUI" and not simply "create"? We don't create widgets here :)

That's right. 😅 The super-class Job has `create()` methods that would clash with this class.
For example if one calls `UIJob.create()` now (without this change) a Job is returned. If we add just a create() method that would now return a `UIJob`. A UIJob is still a Job, so that would not be a compiler-error, but it would suddenly execute in the UI-thread, instead of a worker thread.
One can probably assume that that was the actual intention, nevertheless it would be a behavioral change.

The name createUI is indeed misleading and another would be better (maybe `createUIJob()`?), but just `create()` could cause problems.

laeubi replied:
I don't think you can override static methods anyways...

merks replied:
Indeed you can't, but you can hide the static base class methods by defining a method with the same name in the subclass.  Note however that if one did write UIJob.create that there would be a warning `The static method create(String, ICoreRunnable) from the type Job should be accessed directly` unless one turns off such a warning.   I suppose it's possible that some consumer ignored that warning or turned it off, but that seems somewhat unlikely.

HannesWell replied:
Yes, that's what I meant and yes precisely it is is not a override.

The question to me is if we want to take that risk or not? As you said there is a warning, so one could argue that it is the users fault if that warning is ignored.
And yes, just `UIJob.create()` would definitely be nicer.

merks replied:
I think anyone writing UIJob.create would expect it to create a UIJob and although it doesn't currently do that, changing it now is at worst a source incompatibility not a binary incompatibility.  If someone was affected by the change, they could use Job.create instead as they ought to have done in the first place.  That would still be binary compatible with their intent in both an old Eclipse and a newer one.  So personally I'd be tempted to do what's nicer as an API...

HannesWell replied:
> I think anyone writing UIJob.create would expect it to create a UIJob and although it doesn't currently do that, changing it now is at worst a source incompatibility not a binary incompatibility. 

I think the same. And it is not even a source incompatibility because before UIJob.create() returned a Job and now it returns an `UIJob`, which is still a job. So unless somebody notices less warnings the only (maybe) noticeable effect is that the Job now runs in the UI-thread. That was probably intended in the first place, however it still changed now.

I'm fine with that change. From my experience I just noticed that the Eclipse-community (or maybe consumers in general) is sensitive to changes. Therefore I wanted to raise awareness about this point and if other devs with more experience, like you, say it is OK, I'm fine with that as well. :)

HannesWell commented:
I also noticed there is another, very similar, `UIJob` class in the package `org.eclipse.e4.ui.progress`. I suspect this is because the `org.eclipse.e4.ui.progress` Plug-in should not have a dependency to o.e.workbench.ui?

merks commented:
@HannesWell Yes, I believe you are correct about `org.eclipse.e4.ui.progress` being needed for pure e4 applications that do not have workbench dependencies.

HannesWell commented:
> @HannesWell Yes, I believe you are correct about `org.eclipse.e4.ui.progress` being needed for pure e4 applications that do not have workbench dependencies.

I added the same factories for that UIJob class.
Furthermore I adjusted their run() methods to handle cancellation by throwing `OperationCanceledException` correctly. At least from following the code paths a `OperationCanceledException` would simply lead to an result with error status that has that exception as cause.

HannesWell commented:
Actually I would prefer to have only one factory that accepts a `ICoreRunnable` and ignore the one with `IJobFunction`, but I suspect that this will lead to confusion if only one of two factories from the Job super-class is shadowed. So as long as the Job class has both factory-methods the UIJob should also have both.

@merks, @iloveeclipse are you fine with this change as it is?

iloveeclipse commented on bundles/org.eclipse.ui.navigator/src/org/eclipse/ui/internal/navigator/actions/LinkEditorAction.java:
```diff
@@ -58,61 +55,49 @@ public class LinkEditorAction extends Action implements ISelectionChangedListene
-							});
-							ignoreEditorActivation = false;
+	private UIJob activateEditorJob = UIJob.create(CommonNavigatorMessages.Link_With_Editor_Job_, m -> {
+		LinkEditorAction thisAction = this;
```
`LinkEditorAction thisAction = this;` looks like a hack to avoid compiler error if it wouldn't be a hack.

I would prefer not to refactor **this** code or do it in a way that avoids "hacking"

HannesWell replied:
That's right, without this there is a compiler-error that the fields used below are not yet initialized.
I assume that this is because the lambda is then bound to the fields instead of `this`.
I don't know a better way at the moment and reverted that refactoring for now. If somebody else knows a better way it can be applied in the future.

iloveeclipse commented on bundles/org.eclipse.ui.navigator/src/org/eclipse/ui/internal/navigator/actions/LinkEditorAction.java:
```diff
@@ -58,61 +55,49 @@ public class LinkEditorAction extends Action implements ISelectionChangedListene
+	});
+
+	private UIJob updateSelectionJob = UIJob.create(CommonNavigatorMessages.Link_With_Editor_Job_, m -> {
+		LinkEditorAction thisAction = this;
```
Same here

HannesWell replied:
Reverted, as above.

HannesWell commented:
Does anybody have more remarks/objections? If not I think this is good to go.

I could then also add a N&N entry for this new API.

iloveeclipse reviewed (approved):
LGTM. Please rebase before merge, just in case.

HannesWell commented:
> LGTM. Please rebase before merge, just in case.

Done.

Thank you all.
